[BUG] test_window_aggs_for_rows_lead_lag_on_arrays failed

In last night's EGX-Yarn integration test run, test_window_aggs_for_rows_lead_lag_on_arrays failed:
```
09:52:00  =========================== short test summary info ============================
09:52:00  FAILED integration_tests/src/main/python/window_function_test.py::test_window_aggs_for_rows_lead_lag_on_arrays[partBy:Long-orderBy:Long-orderBy:Long-agg:Array(Double)][IGNORE_ORDER({'local': True})]
09:52:00  FAILED integration_tests/src/main/python/window_function_test.py::test_window_aggs_for_rows_lead_lag_on_arrays[partBy:Long-orderBy:Long-orderBy:Long-agg:Array(Timestamp)][IGNORE_ORDER({'local': True})]
09:52:00  FAILED integration_tests/src/main/python/window_function_test.py::test_window_aggs_for_rows_lead_lag_on_arrays[partBy:Long-orderBy:Long-orderBy:Long-agg:Array(Array(Long))][IGNORE_ORDER({'local': True})]
09:52:00  FAILED integration_tests/src/main/python/window_function_test.py::test_window_aggs_for_rows_lead_lag_on_arrays[partBy:Long-orderBy:Long-orderBy:Long-agg:Array(Array(Array(Double)))][IGNORE_ORDER({'local': True})]
09:52:00  FAILED integration_tests/src/main/python/window_function_test.py::test_window_aggs_for_rows_lead_lag_on_arrays[partBy:Long-orderBy:Long-orderBy:Long-agg:Array(Array(Array(Timestamp)))][IGNORE_ORDER({'local': True})]
```

A sample error log from one of the failures is below.  All of the tests failed with the same, "CPU and GPU list have different lengths" message:
```
09:52:00  =================================== FAILURES ===================================
09:52:00  [31m[1m_ test_window_aggs_for_rows_lead_lag_on_arrays[partBy:Long-orderBy:Long-orderBy:Long-agg:Array(Double)] _[0m
09:52:00  
09:52:00  a_gen = Long, b_gen = Long, c_gen = Long, d_gen = Array(Double)
09:52:00  
09:52:00      @ignore_order(local=True)
09:52:00      @pytest.mark.parametrize('d_gen', lead_lag_array_data_gens, ids=meta_idfn('agg:'))
09:52:00      @pytest.mark.parametrize('c_gen', [long_gen], ids=meta_idfn('orderBy:'))
09:52:00      @pytest.mark.parametrize('b_gen', [long_gen], ids=meta_idfn('orderBy:'))
09:52:00      @pytest.mark.parametrize('a_gen', [long_gen], ids=meta_idfn('partBy:'))
09:52:00      def test_window_aggs_for_rows_lead_lag_on_arrays(a_gen, b_gen, c_gen, d_gen):
09:52:00          data_gen = [
09:52:00                  ('a', RepeatSeqGen(a_gen, length=20)),
09:52:00                  ('b', b_gen),
09:52:00                  ('c', c_gen),
09:52:00                  ('d', d_gen),
09:52:00                  ('d_default', d_gen)]
09:52:00      
09:52:00          assert_gpu_and_cpu_are_equal_sql(
09:52:00              lambda spark: gen_df(spark, data_gen, length=2048),
09:52:00              "window_agg_table",
09:52:00              '''
09:52:00              SELECT
09:52:00                  LEAD(d, 5) OVER (PARTITION by a ORDER BY b,c) lead_d_5,
09:52:00                  LEAD(d, 2, d_default) OVER (PARTITION by a ORDER BY b,c) lead_d_2_default,
09:52:00                  LAG(d, 5) OVER (PARTITION by a ORDER BY b,c) lag_d_5,
09:52:00                  LAG(d, 2, d_default) OVER (PARTITION by a ORDER BY b,c) lag_d_2_default
09:52:00              FROM window_agg_table
09:52:00  >           ''')
09:52:00  
09:52:00  [1m[31mintegration_tests/src/main/python/window_function_test.py[0m:307: 
09:52:00  _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
09:52:00  [1m[31mintegration_tests/src/main/python/asserts.py[0m:400: in assert_gpu_and_cpu_are_equal_sql
09:52:00      assert_gpu_and_cpu_are_equal_collect(do_it_all, conf, is_cpu_first=is_cpu_first)
09:52:00  [1m[31mintegration_tests/src/main/python/asserts.py[0m:364: in assert_gpu_and_cpu_are_equal_collect
09:52:00      _assert_gpu_and_cpu_are_equal(func, 'COLLECT', conf=conf, is_cpu_first=is_cpu_first)
09:52:00  [1m[31mintegration_tests/src/main/python/asserts.py[0m:356: in _assert_gpu_and_cpu_are_equal
09:52:00      assert_equal(from_cpu, from_gpu)
09:52:00  [1m[31mintegration_tests/src/main/python/asserts.py[0m:93: in assert_equal
09:52:00      _assert_equal(cpu, gpu, float_check=get_float_check(), path=[])
09:52:00  [1m[31mintegration_tests/src/main/python/asserts.py[0m:41: in _assert_equal
09:52:00      _assert_equal(cpu[index], gpu[index], float_check, path + [index])
09:52:00  [1m[31mintegration_tests/src/main/python/asserts.py[0m:34: in _assert_equal
09:52:00      _assert_equal(cpu[field], gpu[field], float_check, path + [field])
09:52:00  _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
09:52:00  
09:52:00  cpu = [-7.303240755677303e+32, 1.0843461590776831e-193, -1.2199812804686722e+91]
09:52:00  gpu = [-3.71382200098941e+280, 4.201014304119621e-109, 3.2275294875297787e-196, -2.3707091893006415e-51]
09:52:00  float_check = <function get_float_check.<locals>.<lambda> at 0x7f8866f68d08>
09:52:00  path = [10, 'lag_d_2_default']
09:52:00  
09:52:00      def _assert_equal(cpu, gpu, float_check, path):
09:52:00          t = type(cpu)
09:52:00          if (t is Row):
09:52:00              assert len(cpu) == len(gpu), "CPU and GPU row have different lengths at {} CPU: {} GPU: {}".format(path, len(cpu), len(gpu))
09:52:00              if hasattr(cpu, "__fields__") and hasattr(gpu, "__fields__"):
09:52:00                  for field in cpu.__fields__:
09:52:00                      _assert_equal(cpu[field], gpu[field], float_check, path + [field])
09:52:00              else:
09:52:00                  for index in range(len(cpu)):
09:52:00                      _assert_equal(cpu[index], gpu[index], float_check, path + [index])
09:52:00          elif (t is list):
09:52:00  >           assert len(cpu) == len(gpu), "CPU and GPU list have different lengths at {} CPU: {} GPU: {}".format(path, len(cpu), len(gpu))
09:52:00  [1m[31mE           AssertionError: CPU and GPU list have different lengths at [10, 'lag_d_2_default'] CPU: 3 GPU: 4[0m
09:52:00  
09:52:00  [1m[31mintegration_tests/src/main/python/asserts.py[0m:39: AssertionError
09:52:00  ----------------------------- Captured stdout call -----------------------------
09:52:00  ### CPU RUN ###
09:52:00  ### GPU RUN ###
09:52:00  ### COLLECT: GPU TOOK 0.5071995258331299 CPU TOOK 0.3945033550262451 ###
```

--- a/integration_tests/src/main/python/data_gen.py
+++ b/integration_tests/src/main/python/data_gen.py
@@ -453,14 +453,14 @@ class DateGen(DataGen):
     def __init__(self, start=None, end=None, nullable=True):
         super().__init__(DateType(), nullable=nullable)
         if start is None:
-            # spark supports times starting at
+            # Spark supports times starting at
             # "0001-01-01 00:00:00.000000"
             start = date(1, 1, 1)
         elif not isinstance(start, date):
             raise RuntimeError('Unsupported type passed in for start {}'.format(start))
 
         if end is None:
-            # spark supports time through
+            # Spark supports time through
             # "9999-12-31 23:59:59.999999"
             end = date(9999, 12, 31)
         elif isinstance(end, timedelta):
@@ -513,17 +513,17 @@ class TimestampGen(DataGen):
     def __init__(self, start=None, end=None, nullable=True):
         super().__init__(TimestampType(), nullable=nullable)
         if start is None:
-            # spark supports times starting at
+            # Spark supports times starting at
             # "0001-01-01 00:00:00.000000"
             # but if has issues if you get really close to that because it tries to do things
-            # in a differnt fromat which causes roundoff, so we have to add a few days,
+            # in a different format which causes roundoff, so we have to add a few days,
             # just to be sure
             start = datetime(1, 1, 3, tzinfo=timezone.utc)
         elif not isinstance(start, datetime):
             raise RuntimeError('Unsupported type passed in for start {}'.format(start))
 
         if end is None:
-            # spark supports time through
+            # Spark supports time through
             # "9999-12-31 23:59:59.999999"
             end = datetime(9999, 12, 31, 23, 59, 59, 999999, tzinfo=timezone.utc)
         elif isinstance(end, timedelta):

--- a/integration_tests/src/main/python/data_gen.py
+++ b/integration_tests/src/main/python/data_gen.py
@@ -515,7 +515,10 @@ class TimestampGen(DataGen):
         if start is None:
             # spark supports times starting at
             # "0001-01-01 00:00:00.000000"
-            start = datetime(1, 1, 1, tzinfo=timezone.utc)
+            # but if has issues if you get really close to that because it tries to do things
+            # in a differnt fromat which causes roundoff, so we have to add a few days,
+            # just to be sure
+            start = datetime(1, 1, 3, tzinfo=timezone.utc)
         elif not isinstance(start, datetime):
             raise RuntimeError('Unsupported type passed in for start {}'.format(start))
 

--- a/integration_tests/src/main/python/data_gen.py
+++ b/integration_tests/src/main/python/data_gen.py
@@ -515,7 +515,7 @@ class TimestampGen(DataGen):
         if start is None:
             # Spark supports times starting at
             # "0001-01-01 00:00:00.000000"
-            # but if has issues if you get really close to that because it tries to do things
+            # but it has issues if you get really close to that because it tries to do things
             # in a different format which causes roundoff, so we have to add a few days,
             # just to be sure
             start = datetime(1, 1, 3, tzinfo=timezone.utc)

--- a/integration_tests/src/main/python/window_function_test.py
+++ b/integration_tests/src/main/python/window_function_test.py
@@ -153,13 +153,17 @@ def test_window_aggs_for_ranges_numeric_long_overflow(data_gen):
 
 
 @ignore_order
+@pytest.mark.parametrize('batch_size', ['1000', '1g'], ids=idfn) # set the batch size so we can test multiple stream batches
 @pytest.mark.parametrize('data_gen', [
                                       _grpkey_byte_with_nulls,
                                       _grpkey_short_with_nulls,
                                       _grpkey_int_with_nulls,
                                       _grpkey_long_with_nulls
                                     ], ids=idfn)
-def test_window_aggs_for_range_numeric(data_gen):
+def test_window_aggs_for_range_numeric(data_gen, batch_size):
+    conf = {'spark.rapids.sql.batchSizeBytes': batch_size,
+            'spark.rapids.sql.window.range.byte.enabled': True,
+            'spark.rapids.sql.window.range.short.enabled': True}
     assert_gpu_and_cpu_are_equal_sql(
         lambda spark: gen_df(spark, data_gen, length=2048),
         "window_agg_table",
@@ -192,11 +196,10 @@ def test_window_aggs_for_range_numeric(data_gen):
         '   (partition by a order by b asc  '
         '       range between UNBOUNDED preceding and UNBOUNDED following) as max_b_unbounded '
         'from window_agg_table ',
-        conf={'spark.rapids.sql.window.range.byte.enabled': True,
-              'spark.rapids.sql.window.range.short.enabled': True})
-
+        conf = conf)
 
 @ignore_order
+@pytest.mark.parametrize('batch_size', ['1000', '1g'], ids=idfn) # set the batch size so we can test multiple stream batches
 @pytest.mark.parametrize('data_gen', [_grpkey_longs_with_no_nulls,
                                       _grpkey_longs_with_nulls,
                                       _grpkey_longs_with_timestamps,
@@ -204,7 +207,9 @@ def test_window_aggs_for_range_numeric(data_gen):
                                       _grpkey_longs_with_decimals,
                                       _grpkey_longs_with_nullable_decimals,
                                       _grpkey_decimals_with_nulls], ids=idfn)
-def test_window_aggs_for_rows(data_gen):
+def test_window_aggs_for_rows(data_gen, batch_size):
+    conf = {'spark.rapids.sql.batchSizeBytes': batch_size,
+            'spark.rapids.sql.castFloatToDecimal.enabled': True}
     assert_gpu_and_cpu_are_equal_sql(
         lambda spark : gen_df(spark, data_gen, length=2048),
         "window_agg_table",
@@ -224,7 +229,7 @@ def test_window_aggs_for_rows(data_gen):
         ' row_number() over '
         '   (partition by a order by b,c rows between UNBOUNDED preceding and CURRENT ROW) as row_num '
         'from window_agg_table ',
-        conf = {'spark.rapids.sql.castFloatToDecimal.enabled': True})
+        conf = conf)
 
 
 part_and_order_gens = [long_gen, DoubleGen(no_nans=True, special_cases=[]),
@@ -240,10 +245,13 @@ def meta_idfn(meta):
 
 @ignore_order
 @approximate_float
+@pytest.mark.parametrize('batch_size', ['1000', '1g'], ids=idfn) # set the batch size so we can test multiple stream batches
 @pytest.mark.parametrize('c_gen', lead_lag_data_gens, ids=idfn)
 @pytest.mark.parametrize('b_gen', part_and_order_gens, ids=meta_idfn('orderBy:'))
 @pytest.mark.parametrize('a_gen', part_and_order_gens, ids=meta_idfn('partBy:'))
-def test_multi_types_window_aggs_for_rows_lead_lag(a_gen, b_gen, c_gen):
+def test_multi_types_window_aggs_for_rows_lead_lag(a_gen, b_gen, c_gen, batch_size):
+    conf = {'spark.rapids.sql.batchSizeBytes': batch_size,
+            'spark.rapids.sql.hasNans': False}
     data_gen = [
             ('a', RepeatSeqGen(a_gen, length=20)),
             ('b', b_gen),
@@ -270,7 +278,7 @@ def test_multi_types_window_aggs_for_rows_lead_lag(a_gen, b_gen, c_gen):
                 .withColumn('lag_1_c', f.lag('c', 1).over(baseWindowSpec)) \
                 .withColumn('lag_def_c', f.lag('c', 4, defaultVal).over(baseWindowSpec)) \
                 .withColumn('row_num', f.row_number().over(baseWindowSpec))
-    assert_gpu_and_cpu_are_equal_collect(do_it, conf={'spark.rapids.sql.hasNans': 'false'})
+    assert_gpu_and_cpu_are_equal_collect(do_it, conf = conf)
 
 
 lead_lag_array_data_gens =\
@@ -279,20 +287,14 @@ lead_lag_array_data_gens =\
     [ArrayGen(ArrayGen(ArrayGen(sub_gen, max_length=10), max_length=10), max_length=10) \
         for sub_gen in lead_lag_data_gens]
 
-# lead and lag are supported for arrays, but the other window operations like min and max are not right now
-# once they are all supported the tests should be combined.
-@pytest.mark.skip(reason="If some rows of order-by columns (here is a,b,c) are equal, then it may fail because"
-                          "CPU and GPU can't guarantee the order for the same rows, while lead/lag is typically"
-                          "depending on row's order. The solution is we should add the d and d_default columns"
-                          "into the order-by to guarantee the order. But for now, sorting on array has not been"
-                          "supported yet, see https://github.com/NVIDIA/spark-rapids/issues/2470."
-                          "Once the issue is resolved, we should remove skip mark")
 @ignore_order(local=True)
+@pytest.mark.parametrize('batch_size', ['1000', '1g'], ids=idfn) # set the batch size so we can test multiple stream batches
 @pytest.mark.parametrize('d_gen', lead_lag_array_data_gens, ids=meta_idfn('agg:'))
-@pytest.mark.parametrize('c_gen', [long_gen], ids=meta_idfn('orderBy:'))
+@pytest.mark.parametrize('c_gen', [LongRangeGen()], ids=meta_idfn('orderBy:'))
 @pytest.mark.parametrize('b_gen', [long_gen], ids=meta_idfn('orderBy:'))
 @pytest.mark.parametrize('a_gen', [long_gen], ids=meta_idfn('partBy:'))
-def test_window_aggs_for_rows_lead_lag_on_arrays(a_gen, b_gen, c_gen, d_gen):
+def test_window_aggs_for_rows_lead_lag_on_arrays(a_gen, b_gen, c_gen, d_gen, batch_size):
+    conf = {'spark.rapids.sql.batchSizeBytes': batch_size}
     data_gen = [
             ('a', RepeatSeqGen(a_gen, length=20)),
             ('b', b_gen),
@@ -310,7 +312,8 @@ def test_window_aggs_for_rows_lead_lag_on_arrays(a_gen, b_gen, c_gen, d_gen):
             LAG(d, 5) OVER (PARTITION by a ORDER BY b,c) lag_d_5,
             LAG(d, 2, d_default) OVER (PARTITION by a ORDER BY b,c) lag_d_2_default
         FROM window_agg_table
-        ''')
+        ''',
+        conf = conf)
 
 
 # lead and lag don't currently work for string columns, so redo the tests, but just for strings

